core/capability-negotiation.md

```diff
@@ -0,0 +1,457 @@
+
+If an ACK reply originating from the server is spread across multiple lines, a client MUST NOT
+change capabilities until the last ACK of the set is received. Equally, a server MUST NOT
+change the capabilities of the client until the last ACK of the set has been sent.
+
+### The CAP NAK subcommand
+
+The NAK subcommand designates that the requested capability change was rejected.  The server
+MUST NOT make any change to any capabilities if it replies with a NAK subcommand.
+
+The argument of the NAK subcommand MUST consist of at least the first 100 characters of the
+capability list in the REQ subcommand which triggered the NAK.
+
+Example:
+
+    Client: CAP REQ :multi-prefix sasl ex3
+    Server: CAP * NAK :multi-prefix sasl ex3
+
+### The CAP END subcommand
+
+The END subcommand signals to the server that capability negotiation is complete and requests
+that the server continue with client registration. If the client is already registered, this
+command MUST be ignored by the server.
+
+### The CAP NEW subcommand
+
+The NEW subcommand MUST ONLY be sent to clients that have negotiated CAP Version `302` or
+enabled the `cap-notify` capability.
+
+The NEW subcommand signals that the server supports one or more new capabilities, and may be
+sent at any time. Clients that support `CAP NEW` messages SHOULD respond with a `CAP REQ`
+message if they wish to enable one or more of the newly-offered capabilities.
+
+The format of a `CAP NEW` message is:
+
+    CAP <nick> NEW :<extension 1> [<extension 2> ... [<extension n>]]
+
+As with `LS`, the last parameter is a space-separated list of new capabilities that are now
+offered by the server. If the client supports CAP Version `302`, the capabilities SHOULD be
+listed with values, as in the `CAP LS` response.
+
+Example:
+
+    Server: :irc.example.com CAP modernclient NEW :batch
+
+Example with following `REQ`:
+
+    Server: :irc.example.com CAP tester NEW :away-notify extended-join
+    Client: CAP REQ :extended-join away-notify
+    Server: :irc.example.com CAP tester ACK :extended-join away-notify
+
+### The CAP DEL subcommand
+
+The NEW subcommand MUST ONLY be sent to clients that have negotiated CAP Version `302` or
+enabled the `cap-notify` capability.
+
+The DEL subcommand signals that the server no longer supports one or more capabilities that
+have been advertised. Upon receiving a `CAP DEL` message, the client MUST treat the listed
+capabilities as cancelled and no longer available. Clients SHOULD NOT send `CAP REQ`
+messages to cancel the capabilities in `CAP DEL`, as they have already been cancelled by
+the server.
+
+Servers MUST cancel any capability-specific behavior for a client after sending the
+`CAP DEL` message to the client.
+
+Clients MUST gracefully handle situations when the server removes support for any
+capability. If the client cannot continue to operate without a capability, disconnecting
+with an appropriate `QUIT` message is acceptable.
```

Comment:
When might this happen or be appropriate? Seems a bit weird to be suggesting this, a client is *always* free to disconnect at any time.

Reply:
Yeah this seems weird, I can't think of anything that would trigger this but it feels like there could be some examples. Dunno if it's harmful at all, so I'm fine to keep it in or remove it. Seems to be here more for safety's sake than anything else, "hey clients if you ever do run into this weird case, disconnecting is appropriate"